### PR TITLE
sql: save a backup descriptor next to the data

### PR DESCRIPTION
**Pull request by danhhz**
Use it to discover the locations of files and which one maps to each range.
Store the sql descriptors in it for easy access during restore.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8508)

### CODE REVIEW

**Comment by paperstreet** [comment]
cc @dt

File: sql/backup.go [hunk 1]
```diff
@@ -72,37 +95,47 @@ func Backup(ctx context.Context, db client.DB, base string) (retErr error) {
...
+		return sqlbase.BackupDescriptor{}, err
 	}
 
-	for _, rangeDesc := range rangeDescs {
+	backupDescs := make([]sqlbase.BackupRangeDescriptor, len(rangeDescs))
+	for i, rangeDesc := range rangeDescs {
+		backupDescs[i] = sqlbase.BackupRangeDescriptor{
+			StartKey:  rangeDesc.StartKey.AsRawKey(),
+			EndKey:    rangeDesc.EndKey.AsRawKey(),
+			StartTime: hlc.Timestamp{},
```
Comment (mjibson) [inline sql/backup.go]: What's the point of this field? Can we take it out until we're using it?

**Comment by mjibson** [comment]
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/backup.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-9mr6zuHyCS0yFeRr:-KP-9mr6zuHyCS0yFeRs:b-qch53y) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L87)):*
> ```Go
> func Backup(
> 	ctx context.Context, db client.DB, base string,
> ) (desc sqlbase.BackupDescriptor, retErr error) {
> ```

I can't see anywhere this return is used.

---

*[sql/backup.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-GezwdL95aooG5eHX:-KP-GezwdL95aooG5eHY:be3nvyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L96)):*
> ```Go
> 	txn := client.NewTxn(ctx, db)
> 
> 	rangeDescs, err := allRangeDescriptors(txn)
> ```

Using `retErr` instead of `err` here will work, then the return below can just be `return`. Although maybe not depending on the named return values is good too.

---

*[sql/backup.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-GVt2S24dCYWRAZKm:-KP-GVt3s_-US_tGlzVR:b-kk92ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L98)):*
> ```Go
> 	rangeDescs, err := allRangeDescriptors(txn)
> 	if err != nil {
> 		return sqlbase.BackupDescriptor{}, err
> ```

`return desc, err` here works, too.

---

*[sql/backup.go, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-Fh9sLadJ7jPmzmaW:-KP-Fh9t8le7xMuWheDc:bd8ybdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L111)):*
> ```Go
> 			StartKey:  rangeDesc.StartKey.AsRawKey(),
> 			EndKey:    rangeDesc.EndKey.AsRawKey(),
> 			StartTime: hlc.Timestamp{},
> ```

What's the point of this field? Can we remove it until it's used? And what's the intended difference between this and EndTime below?

---

*[sql/backup.go, line 170 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-H4zoCESI2w_Euwz2:-KP-H4zoCESI2w_Euwz3:b-wkyax8) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L170)):*
> ```Go
> 		return sqlbase.BackupDescriptor{}, err
> 	}
> 	if err = ioutil.WriteFile(filepath.Join(base, backupDescriptorName), descBuf, 0755); err != nil {
> ```

0600 a better permission?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508)*

**Comment by paperstreet** [comment]
RFAL. Added a second commit to make circleci happy

---

Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[sql/backup.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-9mr6zuHyCS0yFeRr:-KP-KKiC4Jf_cv67Tx39:b-1in04u) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L87)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> I can't see anywhere this return is used.
</details>
haha, okay. used it in the cli to print out how many ranges were backed up

---

*[sql/backup.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-GezwdL95aooG5eHX:-KP-KVEUBOOauqM8gnDj:btc6zdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L96)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Using `retErr` instead of `err` here will work, then the return below can just be `return`. Although maybe not depending on the named return values is good too.
</details>
I don't like relying on named return values

---

*[sql/backup.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-GVt2S24dCYWRAZKm:-KP-Kak5UqEIf-_srpu3:b2q6fv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L98)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> `return desc, err` here works, too.
</details>
It does. Is it better?

---

*[sql/backup.go, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-Fh9sLadJ7jPmzmaW:-KP-KeaaTFK0Go8bWrpj:b4hpm8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L111)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> What's the point of this field? Can we remove it until it's used? And what's the intended difference between this and EndTime below?
</details>
It is used to indicate that we backed up every modification between the minimum time and some end (aka a full backup). It'll be non-zero for incremental backups.

Not sure what you mean by the EndTime question? Each range is backed up to be a diff of what's happened between StartTime and EndTime. For restore, we need EndTime to be consistent across the ranges being imported, but StartTime can vary per range if we want to backup some tables more often than others. Consider the following situation:

Backup 1: Table 50 (time [0, 5])
Backup 2: Table 50 (time [5, 10]), Table 51 (time [0, 10])

Then we'll be able to do a full restore of both tables by combining the backups.

---

*[sql/backup.go, line 170 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-H4zoCESI2w_Euwz2:-KP-M5baiQH2SNeI1SJv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L170)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> 0600 a better permission?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-:-KP-N8SF4eWjprknt2wJ:bpk29g5)*

**Comment by mjibson** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/backup.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-GVt2S24dCYWRAZKm:-KP-Pv2Xa6JWbaikDNaA:b7x8c33) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L98)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> It does. Is it better?
</details>
Not sure if it is or not. Leave it as is I guess.

---

*[sql/backup.go, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-Fh9sLadJ7jPmzmaW:-KP-QA7wJpqfNby82rDZ:b-il073d) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L111)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> It is used to indicate that we backed up every modification between the minimum time and some end (aka a full backup). It'll be non-zero for incremental backups.
> 
> Not sure what you mean by the EndTime question? Each range is backed up to be a diff of what's happened between StartTime and EndTime. For restore, we need EndTime to be consistent across the ranges being imported, but StartTime can vary per range if we want to backup some tables more often than others. Consider the following situation:
> 
> Backup 1: Table 50 (time [0, 5])
> Backup 2: Table 50 (time [5, 10]), Table 51 (time [0, 10])
> 
> Then we'll be able to do a full restore of both tables by combining the backups.
</details>
The EndTime field of the BackupDescriptor below that uses txn.Proto.MaxTimestamp. Is StartTime going to differ from that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508)*

**Comment by mjibson** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/backup.go, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-Fh9sLadJ7jPmzmaW:-KP-SOTjMderLnbpTae2:b-rv5as6) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L111)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> The EndTime field of the BackupDescriptor below that uses txn.Proto.MaxTimestamp. Is StartTime going to differ from that?
</details>
Resolved this offline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508)*

**Comment by mjibson** [comment]
:lgtm:

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-:-KP-TGjfKf6OWOao-IVE:bnfp4nl)*

**Comment by mjibson** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/backup.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-9mr6zuHyCS0yFeRr:-KP-TJcMb7cRT0wUK0ju:b9jf6xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L87)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> haha, okay. used it in the cli to print out how many ranges were backed up
</details>
Wait, is this return getting removed or staying?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508)*

**Comment by paperstreet** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/backup.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508#-KP-9mr6zuHyCS0yFeRr:-KP-TUu0-6W4F-07FYC_:b-vqoorg) ([raw file](https://github.com/cockroachdb/cockroach/blob/025304e5889fa9e9b6dce6380b071e19fe66c979/sql/backup.go#L87)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Wait, is this return getting removed or staying?
</details>
staying, it's used now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8508)*

**Pull request merged**